ts/src/main/scala/ch/epfl/ts/traders/BacktestTrader.scala

```diff
@@ -0,0 +1,105 @@
+package ch.epfl.ts.traders
+
+import scala.language.postfixOps
+import scala.collection.mutable.{Map => MMap, MutableList => MList}
+import scala.concurrent.duration.{DurationInt, DurationLong}
+import akka.actor.Cancellable
+import ch.epfl.ts.component.{Component, ComponentRef}
+import ch.epfl.ts.data._
+import ch.epfl.ts.data.Currency._
+
+/**
+  * Implements back testing for traders.
+  *
+  * To use this class, redirect all previous connections into trader to instances
+  * of this class. Connections out of the original trader should remain unchanged.
+  *
+  * @param trader the reference to the real trader component
+  * @param traderId the id of the trader
+  * @param initial the initial seed money
+  * @param currency currency of the intial seed money
+  * @param period the period in milliseconds to compute performance
+  */
+class BacktestTrader(trader: ComponentRef, traderId: Long, initial: Double, currency: Currency, period: Long) extends Component {
+  // for usage of scheduler
+  import context._
+
+  private var schedule: Cancellable = null
+  private val wallet = MMap[Currency, Double](currency -> initial)
+  private val returns = MList[Double]()
+
+  private var lastValue = initial
+  private var maxProfit = 0.0
+  private var maxLoss = 0.0
+
+  def totalReturns: Double = value() / initial * 100
+  def volatility: Double = computeVolatility
+  def drawdown: Double = maxLoss
+  def sharp: Double = totalReturns / volatility
+
+  // handle interested messages and forward all messages to the trader
+  override def receiver = {
+    case t: Transaction if t.buyerId == traderId =>  // buy
+      trader.ar ! t
+      buy(t)
+    case t: Transaction if t.sellerId == traderId =>  // sell
+      trader.ar ! t
+      sell(t)
+    case 'UpdateStatistics =>
+      updateStatistic
+    case m => trader.ar ! m
+  }
+
+  // returns the exchange ratio between two currency
+  private def ratio(from: Currency, to: Currency): Double = ???
```

Comment:
I think you will need to be aware of the current price for the given pair at the time you want to do the conversion.